### ISSUE TITLE
TaskRun with InvalidImageName runs forever

### ISSUE THREAD

**Issue by chmouel** [tektoncd/pipeline#6105]
# Expected Behavior

When an image name of a step is invalid, the taskrun/pipelinerun doesn't get canceled, it waits forever (or as long the timeout is effective)

# Actual Behavior

Fail only until the timeout expired.

# Steps to Reproduce the Problem

1. Create an image with a step and a invalid image (ie: a wrong length sha256)
2. Launch and observe
3.

# Additional Info

Latest pipeline on Latest kind

@lbernick was mentioning the right approach to handle this here:
https://github.com/tektoncd/pipeline/pull/4846#discussion_r877065970

**Comment by pritidesai** [comment]
/assign @vdemeester 

@vdemeester to check if we have addressed this already!

**Comment by pritidesai** [comment]
Pipelines WG - @jerop please move to the next milestone if we haven't heard back on this before releasing 0.48

**Comment by JeromeJu** [comment]
WG
/assign @afrittoli

**Comment by afrittoli** [comment]
I left a comment here as a start https://github.com/tektoncd/pipeline/pull/4846#discussion_r1233137854 - I will check if there is a way we can make the `Pod` fail right away in case of `InvalidImageName`

**Comment by afrittoli** [comment]
@chmouel I would argue that this is not critical urgent. A faster fail would be nice to have, however, if I understood correctly, the k8s Pod controller will keep trying until a timeout, so Tekton honours that behaviour.

**Comment by afrittoli** [comment]
The `TaskRun` controller reviews the `Pod` status, and when its phase is `Pending` it updates the `TaskRun` accordingly to a similar pending state. Marking the `TaskRun` as failed when the `InvalidImageName` is found is easy, the problem though is that that would lead to a failed `TaskRun` with a pending `Pod` which could lead to an inconsistent case.

Is there a way other than deletion to terminate the pod? 
One way could be to rewrite the pod spec to something that fails, but that might be confusing to the end user.

Assuming a termination message can be sent to the `Pod`, after that there are two options:
- set the `TaskRun` status to failed, and trust that the `Pod` will stop eventually. When that happens, if the Pod is marked as failed and the `TaskRun` was already failed, ignore the update
- leave the `TaskRun` status pending. When the `Pod` processes the termination and stops, in the associated reconcile cycle update the `TaskRun` to failed. It may be difficult in this scenario to carry over the real reason for failure, I need to check if the termination can be associated with a message

**Comment by afrittoli** [comment]
@chmouel @tektoncd/core-maintainers it seems to me that a good solution is more complex than the value it brings, but I'd be happy to continue working on this if you think it's worth it. We need to decide what the behaviour of the `Pod` will be though. One way to handle this with current logic would be to cancel the taskrun when an invalid image is detected.

**Comment by lbernick** [comment]
@afrittoli thank you for looking into this! I think deleting the pod and failing the TaskRun makes the most sense. I don't see a problem with deletion if the image pull error isn't retryable (hopefully we have a way of determining if the error is retryable or not).

Cancelation might be a bit confusing but I think it would also be an OK solution. I just want to note that if we choose cancelation for invalid image pulls and end up implementing support for canceling the taskrun via the entrypoint (https://github.com/tektoncd/pipeline/issues/3238), we may need to replace the user-provided image with a noop image or something like that, since the entrypoint can't start before the image is pulled.

**Comment by vdemeester** [comment]
@afrittoli I do also think deleting the pod and failing the taskrun makes the most sense. We just need to make sure we provide enough information to the users in the TaskRun for them not to be confused that there is not `Pod` attached.
